app/src/main/kotlin/com/looker/droidify/utility/InstallUtils.kt

```diff
@@ -0,0 +1,62 @@
+package com.looker.droidify.utility
+
+import android.content.Context
+import com.looker.core.common.PackageName
+import com.looker.core.datastore.SettingsRepository
+import com.looker.core.domain.InstalledItem
+import com.looker.core.domain.Product
+import com.looker.core.domain.Repository
+import com.looker.droidify.service.Connection
+import com.looker.droidify.service.DownloadService
+import com.looker.droidify.utility.extension.startUpdate
+import com.looker.installer.InstallManager
+import com.looker.installer.model.InstallItem
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+class InstallUtils {
+
+    companion object {
+        fun install(
+            context: Context,
+            settingsRepository: SettingsRepository,
+            scope: CoroutineScope,
+            installedItem: InstalledItem,
+            products: List<Pair<Product, Repository>>
+        ) {
+            Connection(
+                serviceClass = DownloadService::class.java,
+                onBind = { _, binder ->
+                    scope.launch {
+
+                        val state = binder.downloadState.value
+
+                        if (state.currentItem is DownloadService.State.Success) {
+
+                            InstallManager(context, settingsRepository)
+                                .install(
+                                    InstallItem(
+                                        PackageName(state.currentItem.packageName),
+                                        state.currentItem.release.cacheFileName
+                                    )
+                                )
+                        }
+                    }
+                }
+            ).startUpdate(installedItem.packageName, installedItem, products)
+        }
+
+        fun uninstall(
+            context: Context,
+            settingsRepository: SettingsRepository,
+            scope: CoroutineScope,
+            packageName: PackageName,
+        ) {
+            scope.launch {
+                InstallManager(context, settingsRepository)
```

Comment:
@Iamlooker `InstallManager#uninstall` gets called, but `uninstallItems.consumeEach` doesn't get triggered. why? 🤔 

PS: sorry for the delay. had exams, getting back to coding only now!

Reply:
Instead what you can do is, add this to viewModel,

```
fun uninstallPackage(packageName: String) {
        viewModelScope.launch {
            installManager.uninstall(packageName.toPackageName())
        }
    }
```

As creating a new instance of InstallManager is very bad, because it is a heavy class. I think a new instance of InstallManager is causing a race situation

Reply:
Also I suggest doing this in view model for install:

Create a download connection like this:

```
val downloadConnection = Connection(
        serviceClass = DownloadService::class.java,
        onBind = { _, binder ->
            viewModelScope.launch {

                val state = binder.downloadState.value

                if (state.currentItem is DownloadService.State.Success) {

                    installManager
                        .install(
                            state.currentItem.packageName installFrom state.currentItem.release.cacheFileName
                        )
                }
            }
        }
    )
```

Bind it like the sync connection

and create a install function like this:

```
fun install(packageName: String, products: List<Pair<Product, Repository>>) {
        viewModelScope.launch {
            val installedItem = Database.InstalledAdapter.get(packageName, null)
            downloadConnection.startUpdate(packageName, installedItem, products)
        }
    }
```

This will remove the need for InstallUtils object and also fix the issue